### ISSUE TITLE
Unusable performance with >1k items in ClusterManager

### ISSUE THREAD

**Issue by mbilalhussain96** [googlemaps/android-maps-utils#1199]
If I add over 1k items to the ClusterManager, the actual (drawn) markers are updated after very long delays. For example, I can have one cluster with "1000+" text on it, and when I zoom in, it stays for about 1-2 minutes, and only after that it splits into smaller clusters and markers.

```
  private lateinit var mMap: GoogleMap, 
   private lateinit var clusterManager: ClusterManager<mapclass>

  private fun addItems() {

        // welcomedataItemList1=1000

        for (place in 0 until welcomedataItemList1.size) {

            if (welcomedataItemList1[place]?.lat != null && welcomedataItemList1[place]?.lon != null) {

                var latitude =welcomedataItemList1[place].lat.toDouble()
                var longitude =welcomedataItemList1[place].lon.toDouble()
                serial = welcomedataItemList1[place].serial
                ownerName = welcomedataItemList1[place].ownerName
              
                status= welcomedataItemList1[place].adminStatus.toInt()
                mMap.moveCamera(CameraUpdateFactory.newLatLngZoom(LatLng(latitude, longitude), 6f))
                var offsetItem = mapclass(latitude, longitude, "$serial", "$ownerName")
                clusterManager.addItem(offsetItem)

            }
        }

    }


private fun setUpClusterer() {

        val metrics = DisplayMetrics()
        windowManager.defaultDisplay.getMetrics(metrics)

        clusterManager = ClusterManager(this, mMap)
        clusterManager.setAlgorithm(NonHierarchicalViewBasedAlgorithm(metrics.widthPixels, metrics.heightPixels))
        mMap.setOnCameraIdleListener(clusterManager)



        addItems()

        clusterManager.renderer = MarkerClusterRenderer(this, mMap, clusterManager)

        mMap.setInfoWindowAdapter(CustomInfoWindowForGoogleMap(this))

        mMap.setOnMarkerClickListener(clusterManager)
        mMap.uiSettings.isZoomControlsEnabled= true
        clusterManager.setAnimation(false);
        clusterManager.cluster()

    }


override fun onMapReady(googleMap: GoogleMap) {

      mMap = googleMap
            setUpClusterer()
        
        }
```

**Comment by wangela** [comment]
If you would like to upvote the priority of this issue, please comment below or react with :+1: so we can see what is popular when we triage.
--------

@mbilalhussain96 Thank you for opening this issue. 🙏
Please check out these other resources that might help you get to a resolution in the meantime:

- [Check the issue tracker](https://developers.google.com/maps/support/#issue_tracker) - bugs and feature requests for Google Maps Platform APIs and SDKs
- [Open a support case](https://developers.google.com/maps/support#contact-maps-support) - Get 1:1 support in Cloud Console.
- [Discord](https://discord.gg/fHTDmZe9En) - chat with other developers
- [StackOverflow](http://stackoverflow.com/questions/tagged/google-maps) - use the `google-maps` tag


_This is an automated message, feel free to ignore._

**Comment by wangela** [comment]
A suggestion from @cwsiteplan in the [Compose issue discussion](https://github.com/googlemaps/android-maps-compose/issues/380#issuecomment-1784821181):

> would it be possible to expand clusters only for the visible viewport? - seems like it's currently expanding on zoom level - adding lot's of individual markers on areas that are potentially not viewed at all

**Comment by ColtonIdle** [comment]
Hm. so if we care about performance. we shouldn't use compose for clustering?

**Comment by ColtonIdle** [comment]
Yeah, I currently have an app where we have about 16k items (mostly US and europe) hence my curiosity because we're not (yet) using compose maps. but we are using a fully compose android app with the exception of the map =)

im really eager to ditch AndroidView though as I've had a bunch of issues with clusters not recomposing properly. thanks for the update. and strong skipping seems cool. i wonder if we can just mark latLng as @Stable though?

**Comment by ShikaSD** [comment]
Hi, one of the Compose devs here :)

Could you share a bit more about performance problems you have encountered? I am not personally familiar with Compose integration for Maps, but I'd be happy to chat about it and help debug / address perf issues with it.

**Comment by ShikaSD** [comment]
@kikoso stability by itself does not result in performance problems that often, I doubt this is the cause here. I suspect the problem is in heavy content used for clusters here. I'll look into attached issue a bit more

**Comment by wangela** [comment]
Can anyone verify that using `NonHierarchicalViewBasedAlgorithm` resolves your high-volume clustering performance issues? @mbilalhussain96 I see in your original post that you were already specifying `NonHierarchicalViewBasedAlgorithm`, but your specifications for width and height differ from the [demo app](https://github.com/googlemaps/android-maps-utils/blob/main/demo/src/main/java/com/google/maps/android/utils/demo/VisibleClusteringDemoActivity.java#L49) which specifies `metrics.widthPixels / metrics.density` for width and `metrics.heightPixels / metrics.density` for height.

cc @ColtonIdle

**Comment by Willkxxx** [comment]
Listen I have never used this site I don’t know how honestly I want to
report fraud or someone has hacked my account delete this account

**Comment by ColtonIdle** [comment]
For anyone that was curious of how to grab height and width from compose code, you can see a discussion I brought up in kotlinlang slack: https://slack-chats.kotlinlang.org/t/16243087/what-would-be-the-compose-way-of-grabbing-these-window-value#eb72b9b4-d4a9-462d-bbd7-c727dfed29d3

**Comment by LouisCAD** [comment]
TL;DR is use `Modifier.onSizeChanged`.
